File: python/301-350/329. Longest Increasing Path in a Matrix.py
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        def dfs(i, j):
            if not dp[i][j]:
                val = matrix[i][j]
                dp[i][j] = 1 + max(
                    dfs(i-1,j) if i and val > matrix[i-1][j] else 0,
                    dfs(i+1,j) if i+1 < len(matrix) and val > matrix[i+1][j] else 0,
                    dfs(i, j - 1) if j-1 >= 0 and val > matrix[i][j-1] else 0,
                    dfs(i, j + 1) if j + 1 < len(matrix[0]) and val > matrix[i][j+1] else 0)
            return dp[i][j]
        if not matrix :
            return 0
        dp = [[0] * len(matrix[0]) for _ in range(len(matrix))]
        return max(dfs(x,y) for x in range(len(matrix)) for y in range(len(matrix[0])))

File: cpp/301-350/329. Longest Increasing Path in a Matrix.cc
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int res = 0;
        if(matrix.empty())
        	return res;
        vector<vector<int>> visited(matrix.size(), vector<int>(matrix[0].size(), 0));
        for (int i = 0; i < matrix.size(); i++) {
            for (int j = 0; j < matrix[0].size(); j++) {
                res = max(res, dfs(i, j, INT_MIN, 0, res, matrix, visited));
            }
        }
        return res;
    }
    int dfs(int i, int j, int lastval, int length, int &res, vector<vector<int>> &matrix, vector<vector<int>> &visited) {
        if (i < 0 || j < 0 || i >= matrix.size() || j >= matrix[0].size() || matrix[i][j] <= lastval) {
            return 0;
        }
        if(visited[i][j])
        	return visited[i][j];
        int pos[] = {0, 1, 0, -1, 0};
        int maxval = 0;
        for (int k = 0; k < 4; k++) {
            int x = i + pos[k];
            int y = j + pos[k + 1];
            maxval = max(maxval, dfs(x, y, matrix[i][j], length + 1, res, matrix, visited));
        }
        visited[i][j] = maxval + 1;
        return visited[i][j];
    }
};